### PR TITLE
#1959. Tests for class modifiers restrictions and type aliases

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t23.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// - A declaration has a base or final superdeclaration, and is not itself
+///   marked base, final or sealed.
+///
+/// @description Check that it is a compile-time error if a declaration is not
+/// `base`, `final` or `sealed` and has a superdeclaration marked `final`. Test
```
Comment (chloestefantsova) [inline LanguageFeatures/Class-modifiers/basic_restrictions_A04_t23.dart:11]: If I understood the spec correctly, a declaration that has a proper superdeclaration marked `final` is a compile-time error regardless of the declaration itself having `base`, `final`, or `sealed` modifiers. Some exceptions may apply, based on the version of the libraries and them being core libraries.

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t24.dart [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// - A declaration has a base or final superdeclaration, and is not itself
+///   marked base, final or sealed.
+///
+/// @description Check that it is a compile-time error if a declaration is not
+/// `base`, `final` or `sealed` and has a superdeclaration marked `final`. Test
```
Comment (chloestefantsova) [inline LanguageFeatures/Class-modifiers/basic_restrictions_A04_t24.dart:11]: Ditto.

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t25.dart [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// - A declaration has a base or final superdeclaration, and is not itself
+///   marked base, final or sealed.
+///
+/// @description Check that it is a compile-time error if a declaration is not
+/// `base`, `final` or `sealed` and has a superdeclaration marked `final`. Test
```
Comment (chloestefantsova) [inline LanguageFeatures/Class-modifiers/basic_restrictions_A04_t25.dart:11]: Ditto.

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t26.dart [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// - A declaration has a base or final superdeclaration, and is not itself
+///   marked base, final or sealed.
+///
+/// @description Check that it is a compile-time error if a declaration is not
+/// `base`, `final` or `sealed` and has a superdeclaration marked `final`. Test
```
Comment (chloestefantsova) [inline LanguageFeatures/Class-modifiers/basic_restrictions_A04_t26.dart:11]: Ditto.

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t27.dart [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// - A declaration has a base or final superdeclaration, and is not itself
+///   marked base, final or sealed.
+///
+/// @description Check that it is a compile-time error if a declaration is not
+/// `base`, `final` or `sealed` and has a superdeclaration marked `sealed`. Test
```
Comment (chloestefantsova) [inline LanguageFeatures/Class-modifiers/basic_restrictions_A04_t27.dart:11]: If I understood the specification correctly, the restriction imposed by the `sealed` declarations is that they can be extended only in the same library they are declared. It is a compile-time error to use them as superdeclarations outside of their own library, regardless of the modifiers `base`, `final`, and `sealed`.

**Review by chloestefantsova — CHANGES_REQUESTED** [review]

File: LanguageFeatures/Class-modifiers/basic_restrictions_A01_t11.dart [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+// [cfe] unspecified
+
+enum EnumWithSealedClass with TypedefSealedClass {e1, e2}
+//                            ^^^^^^^^^^^^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+
+typedef LocalTypedefSealedClass = SealedClass;
+
+mixin MixinOnLocalTypedefSealedClasson LocalTypedefSealedClass {}
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Class-modifiers/basic_restrictions_A01_t11.dart]: Thank you. Fixed

File: LanguageFeatures/Class-modifiers/basic_restrictions_A03_t19.dart [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+// [analyzer] unspecified
+// [cfe] unspecified
+
+base mixin MixinIndirectlyImplementsBaseClass implements TypedefWithBaseMixinClass {}
+//                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+
+base mixin class MixinClassIndirectlyImplementsBaseClass implements TypedefWithBaseMixinClass {}
+//                                                                  ^^^^^^^^^^^^^^^^^^
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Class-modifiers/basic_restrictions_A03_t19.dart]: Fixed

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t23.dart (hunk 1 above)
Comment (sgrekhov) [reply LanguageFeatures/Class-modifiers/basic_restrictions_A04_t23.dart:11]: `final` class can be extended by `final`, `base` or `sealed` classes defined in the same library. One of the tests that checks it (and passes) is https://github.com/dart-lang/co19/blob/master/LanguageFeatures/Class-modifiers/basic_restrictions_A02_t07.dart

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t27.dart (hunk 5 above)
Comment (sgrekhov) [reply LanguageFeatures/Class-modifiers/basic_restrictions_A04_t27.dart]: Indeed. And we have `basic_restrictions_A01_t01.dart` (`basic_restrictions_A01_t09.dart` for type aliases) testsing that  it's an error to extend a `sealed` class outside of its library by a class with any modifier. Deleted this test. Thank you for noticing

**Review by sgrekhov — COMMENTED** [review]
Typos fixed, wrong tests deleted. Please take another look

File: LanguageFeatures/Class-modifiers/basic_restrictions_A04_t23.dart (hunk 1 above)
Comment (chloestefantsova) [reply LanguageFeatures/Class-modifiers/basic_restrictions_A04_t23.dart:11]: Got it. Thanks for the clarification!

**Review by chloestefantsova — APPROVED** [review]
LGTM. Thanks!

**Comment by sgrekhov** [comment]
Merge conflicts resolved. Please merge this PR

**Pull request merged**